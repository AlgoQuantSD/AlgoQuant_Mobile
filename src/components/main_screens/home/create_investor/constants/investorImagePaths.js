export const NUM_INVESTOR_IMAGES_PER_FREQ = 3;

export const INVESTOR_IMAGE_BASE_URL =
  "https://algoquant-resources.s3.amazonaws.com/InvestorImages";
