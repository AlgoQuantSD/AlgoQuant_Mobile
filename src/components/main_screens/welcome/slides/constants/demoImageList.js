import DemoInvestorImage0 from "../../../../../../assets/demo_images/investor-0.png";
import DemoInvestorImage1 from "../../../../../../assets/demo_images/investor-1.png";
import DemoInvestorImage2 from "../../../../../../assets/demo_images/investor-2.png";
import DemoInvestorImage3 from "../../../../../../assets/demo_images/investor-3.png";
import DemoInvestorImage4 from "../../../../../../assets/demo_images/investor-4.png";
import DemoInvestorImage5 from "../../../../../../assets/demo_images/investor-5.png";
import DemoInvestorImage6 from "../../../../../../assets/demo_images/investor-6.png";

// List of images to display in the demo carousel
export const DEMO_IMAGES_INVESTOR = [
  {
    image: DemoInvestorImage0,
    caption:
      "Begin by pressing the Create Investor button in the bottom right corner of the screen.",
    id: 0,
  },
  {
    image: DemoInvestorImage1,
    caption: "Name your investor and select an investor type.",
    id: 1,
  },
  {
    image: DemoInvestorImage2,
    caption: "Select indicators and a trade frequency.",
    id: 2,
  },
  {
    image: DemoInvestorImage3,
    caption: "Learn more about indicators by pressing the tooltips.",
    id: 3,
  },
  {
    image: DemoInvestorImage4,
    caption:
      "Profit and loss stop dictates the threshold that must be met in order to cash out.",
    id: 4,
  },
  {
    image: DemoInvestorImage5,
    caption: "Select stock for your investor to consider.",
    id: 5,
  },
  {
    image: DemoInvestorImage6,
    caption:
      "Make sure everything looks good before finalizing the creation process.",
    id: 6,
  },
];
