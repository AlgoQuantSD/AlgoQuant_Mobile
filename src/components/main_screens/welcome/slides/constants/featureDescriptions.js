export const INVESTOR_DESCRIPTION = [
  { text: "Customize your investor to your needs", id: 0 },
  { text: "Use technical analysis or AI to configure your investor", id: 1 },
  { text: "Select what assets you want your investor to track", id: 2 },
  { text: "Create jobs for your investor", id: 3 },
];

export const BACKTEST_DESCRIPTION = [
  { text: "Test your investor with historical market data", id: 0 },
  {
    text: "Compare your investor's performance against typical buy/hold strategies",
    id: 1,
  },
  { text: "View in depth statistics about your investor's performance", id: 2 },
  { text: "Gain confidence in your investor", id: 3 },
];

export const JOB_DESCRIPTION = [
  { text: "Deploy your investor into the market via a job", id: 0 },
  {
    text: "Allocate money to your jobs",
    id: 1,
  },
  { text: "View your job's performance in real time", id: 2 },
];

export const PAPER_TRADING_DESCRIPTION = [
  {
    text: "Connect your Alpaca account via API keys for a realistic paper trading experience",
    id: 0,
  },
  {
    text: "Utilize the AlgoQuant simulated brokerage if you do not wish to connect your Alpaca account",
    id: 1,
  },
  { text: "Monitor your portfolio balance", id: 2 },
];
