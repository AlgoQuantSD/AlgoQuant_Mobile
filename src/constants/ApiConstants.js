import React from "react";

export const TRADE_HISTORY_FETCH_AMOUNT = 10;