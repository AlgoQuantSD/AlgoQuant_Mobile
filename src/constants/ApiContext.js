import { AlgoQuant } from "algoquant/lib/AlgoQuant";
import React from "react";

const algoqant = new AlgoQuant();
const AlgoquantApiContext = React.createContext(algoqant);
export default AlgoquantApiContext;
