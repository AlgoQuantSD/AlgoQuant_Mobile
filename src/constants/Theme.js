const Palette = {
  primaryGreen: "#3F9F30",
  primaryGreenOpacity35: "#3F9F3059",
  red: "#B11818",
  lightGray: "#C7C7C7",
  mediumGray: "#747272",
  darkGray: "#323232",
  darkGrayOpacity95: "#323232F2",
  white: "#FFFFFF",
  whiteOpacity35: "FFFFFFE6",
  black: "#000000",
};

export const Theme = {
  colors: {
    background: Palette.black,
    foreground: Palette.white,
    primary: Palette.green,
    success: Palette.green,
    danger: Palette.red,
  },
  text: {
    fontSizeH1: 28,
    fontSizeBody: 16,
    fontSizeButton: 16,
    color: Palette.white,
    disabledColor: Palette.mediumGray,
  },
  button: {
    alignItems: "center",
    justifyContent: "center",
    width: 120,
    height: 45,
    marginHorizontal: 20,
    borderRadius: 10,
    padding: 3,
    backgroundColor: Palette.primaryGreen,
    disabledBackgroundColor: Palette.primaryGreenOpacity35,
  },
  modal: {
    backgroundColor: Palette.darkGrayOpacity95,
  },
};

export const LightTheme = {
  ...Theme,
  colors: {
    background: Palette.white,
    foreground: Palette.black,
    text: Palette.black,
  },
};
