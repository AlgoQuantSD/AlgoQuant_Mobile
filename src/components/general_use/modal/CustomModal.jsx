import { Ionicons } from "@expo/vector-icons";
import React, { useContext, useEffect, useState } from "react";
import {
  ActivityIndicator,
  Keyboard,
  StyleSheet,
  Text,
  TouchableOpacity,
  TouchableWithoutFeedback,
  View,
} from "react-native";
import Modal from "react-native-modal";
import { Button, Snackbar, TextInput } from "react-native-paper";
import AlgoquantApiContext from "../../../constants/ApiContext";
import { THEME } from "../../../constants/Theme";
import {
  submitConnectAlpacaModal,
  submitDeleteAccountModal,
  submitDeleteInvestorModal,
  submitDisconnectAlpacaModal,
  submitEditNameModal,
  submitResetBalanceModal,
  submitResetPasswordModal,
  submitStartJobModal,
  submitStopJobModal,
  submitUpdateEmailConfirmNewEmailStep,
  submitUpdateEmailModalNewEmailStep,
  submitUpdatePhoneModal,
} from "../../../helpers/modalSubmitActions";
import { snackbarCleanUp } from "../../../helpers/snackbarCleanup";

export default function CustomModal(props) {
  // Get the information for the modal and the functions to clear the information when we close the modal
  const {
    isModalVisible,
    setIsModalVisible,
    modalType,
    setModalType,
    modalTitle,
    setModalTitle,
    modalHeader,
    setModalHeader,
    modalBody,
    setModalBody,
    modalInputFields,
    setModalInputFields,
    modalButtons,
    setModalButtons,
    snackbarMessage,
    setSnackbarMessage,
    isSnackbarVisible,
    setIsSnackbarVisible,
    modalSnackbarMessage,
    setModalSnackbarMessage,
    isModalSnackbarVisible,
    setIsModalSnackbarVisible,
    modalErrorMessage,
    setModalErrorMessage,
    setInvestorListRefresh,
    jobID,
    investorID,
    navigation,
  } = props;

  // Get the API context for the modals that require it
  const algoquantApi = useContext(AlgoquantApiContext);

  // Close the modal and clear the modal information
  function handleModalClose() {
    setIsModalVisible(!isModalVisible);
    setModalType(null);
    setModalTitle(null);
    setModalHeader(null);
    setModalBody(null);
    if (setModalInputFields) {
      setModalInputFields(null);
    }
    setModalButtons(null);
    if (setInputValues) {
      setInputValues(null);
    }
    if (setIsModalSnackbarVisible) {
      setIsModalSnackbarVisible(false);
    }
    if (setModalSnackbarMessage) {
      setModalSnackbarMessage(null);
    }
    if (setIsSensitiveTextHidden) {
      setIsSensitiveTextHidden(true);
    }
  }

  // Perform the appropriate action upon submitting based on which modal is open
  function handleSubmit() {
    let submitProps = {
      inputValues,
      modalType,
      setInputValues,
      modalInputFields,
      isModalVisible,
      setIsModalVisible,
      setModalType,
      setModalTitle,
      setModalHeader,
      setModalBody,
      setModalInputFields,
      setModalButtons,
      isSnackbarVisible,
      setIsSnackbarVisible,
      snackbarMessage,
      setSnackbarMessage,
      modalSnackbarMessage,
      setModalSnackbarMessage,
      isModalSnackbarVisible,
      setIsModalSnackbarVisible,
      modalErrorMessage,
      setModalErrorMessage,
      isLoading,
      setIsLoading,
      setInvestorListRefresh,
      jobID,
      investorID,
      navigation,
      algoquantApi,
    };
    switch (modalType) {
      case "EDIT_NAME":
        submitEditNameModal(submitProps);
        break;
      case "RESET_ALPACA_BALANCE":
      case "RESET_SIMULATED_BALANCE":
        submitResetBalanceModal(submitProps);
        break;
      case "CONNECT_ALPACA":
        submitConnectAlpacaModal(submitProps);
        break;
      case "DISCONNECT_ALPACA":
        submitDisconnectAlpacaModal(submitProps);
        break;
      case "RESET_PASSWORD":
        submitResetPasswordModal(submitProps);
        break;
      case "UPDATE_EMAIL_NEW_EMAIL_STEP":
        submitUpdateEmailModalNewEmailStep(submitProps);
        break;
      case "UPDATE_EMAIL_NEW_EMAIL_CONFIRM_STEP":
        submitUpdateEmailConfirmNewEmailStep(submitProps);
        break;
      case "UPDATE_PHONE":
        submitUpdatePhoneModal(submitProps);
        break;
      case "DELETE_ACCOUNT":
        submitDeleteAccountModal(submitProps);
        break;
      case "DELETE_INVESTOR":
        submitDeleteInvestorModal(submitProps);
        break;
      case "START_JOB":
        submitStartJobModal(submitProps);
        break;
      case "STOP_JOB":
        submitStopJobModal(submitProps);
        break;
      default:
    }
  }

  // Keep track of what is being typed into the input fields if we have any in our modal
  const [inputValues, setInputValues] = useState(
    Array(modalInputFields?.length).fill("")
  );
  const [isLoading, setIsLoading] = useState(false);

  const [isSensitiveTextHidden, setIsSensitiveTextHidden] = useState(true);
  // Refresh the component to get the correct amount of input values
  useEffect(() => {
    if (modalInputFields) {
      setInputValues(Array(modalInputFields.length).fill(""));
    }
  }, [modalInputFields]);

  // Check whether certain pieces of the modal should be rendered if so we render it
  // In some cases such as input fields and buttons we must loop through and render them because the number of input fields and buttons in a modal may vary
  return (
    <TouchableWithoutFeedback onPress={() => Keyboard.dismiss()}>
      <Modal
        isVisible={isModalVisible}
        onBackdropPress={handleModalClose}
        style={styles.modalContainer}
      >
        {isLoading ? (
          <View style={styles.loadingIndicator}>
            <ActivityIndicator
              size="large"
              color={THEME.activityIndicator.color.secondary}
            />
          </View>
        ) : null}
        <View style={styles.headerContainer}>
          <Text
            adjustsFontSizeToFit={true}
            numberOfLines={1}
            style={styles.titleText}
          >
            {modalTitle}
          </Text>
          <TouchableOpacity
            style={styles.closeButton}
            onPress={handleModalClose}
          >
            <Ionicons name="close" size={16} color="white" />
          </TouchableOpacity>
        </View>
        <View style={styles.contentContainer}>
          {modalHeader ? (
            <View style={styles.modalHeader} testID="modal-header">
              <Text style={styles.headerText}>{modalHeader}</Text>
            </View>
          ) : null}
          {modalBody ? (
            <View style={styles.modalBody} testID="modal-body">
              <Text style={styles.bodyText}>{modalBody}</Text>
            </View>
          ) : null}
          {modalInputFields ? (
            <View style={styles.modalInputFields} testID="modal-input-fields">
              {modalInputFields.map((item, index) => (
                <View
                  key={item.key}
                  style={{
                    flexDirection: "row",
                    width: "100%",
                  }}
                >
                  <TextInput
                    label={item.label}
                    secureTextEntry={
                      modalType === "RESET_PASSWORD"
                        ? isSensitiveTextHidden
                        : false
                    }
                    keyboardType={
                      item.keyboardType === "number-pad"
                        ? "number-pad"
                        : "default"
                    }
                    defaultValue={item.defaultValue}
                    onChangeText={(text) => {
                      const newInputValues = [...inputValues];
                      newInputValues[index] = text;
                      setInputValues(newInputValues);
                    }}
                    selectionColor={THEME.colors.background}
                    underlineColor={THEME.colors.background}
                    activeUnderlineColor={THEME.colors.background}
                    outlineColor={THEME.colors.background}
                    activeOutlineColor={THEME.colors.background}
                    textColor={THEME.colors.background}
                    placeholderTextColor={THEME.colors.background}
                    contentStyle={{ color: THEME.colors.background }}
                    style={{
                      backgroundColor: THEME.colors.transparent,
                      width: "100%",
                    }}
                  />
                  {modalType === "RESET_PASSWORD" ? (
                    <TouchableOpacity
                      style={{
                        alignSelf: "center",
                        position: "absolute",
                        left: "90%",
                      }}
                      onPress={() =>
                        setIsSensitiveTextHidden(!isSensitiveTextHidden)
                      }
                    >
                      <Ionicons
                        name="eye"
                        size={20}
                        color={THEME.colors.foreground}
                      />
                    </TouchableOpacity>
                  ) : null}
                </View>
              ))}
            </View>
          ) : null}
        </View>
        {modalButtons ? (
          <View style={styles.modalButtons}>
            {modalButtons.map((item, index) => (
              <TouchableOpacity key={item.key} style={styles.button}>
                <Button
                  onPress={
                    item.key === "SUBMIT_BUTTON"
                      ? handleSubmit
                      : handleModalClose
                  }
                  children={item.label}
                  buttonColor={item.buttonColor}
                  textColor={item.textColor}
                  style={THEME.button.style}
                />
              </TouchableOpacity>
            ))}
          </View>
        ) : null}
        <View
          style={{
            position: "absolute",
            bottom: -150,
            width: "100%",
          }}
        >
          <Snackbar
            visible={isModalSnackbarVisible}
            onDismiss={() =>
              snackbarCleanUp(
                setIsModalSnackbarVisible,
                setModalSnackbarMessage
              )
            }
            duration={3500}
            action={{
              label: "Dismiss",
              textColor: THEME.text.color.secondary,
              onPress: () => {
                snackbarCleanUp(
                  setIsModalSnackbarVisible,
                  setModalSnackbarMessage
                );
              },
            }}
            style={styles.snackbar}
          >
            {modalSnackbarMessage}
          </Snackbar>
        </View>
      </Modal>
    </TouchableWithoutFeedback>
  );
}

const styles = StyleSheet.create({
  modalContainer: {
    justifyContent: "flex-start",
    alignItems: "center",
    backgroundColor: THEME.modal.color.background,
    flex: 0.75,
    marginTop: "50%",
    borderRadius: 10,
  },
  contentContainer: {
    flex: 0.6,
    alignItems: "center",
    justifyContent: "flex-start",
    width: "90%",
  },
  headerContainer: {
    position: "relative",
    flex: 0.15,
    flexDirection: "row",
    justifyContent: "center",
    alignItems: "center",
    width: "100%",
  },
  titleText: {
    color: THEME.text.color.secondary,
    fontWeight: "bold",
    fontSize: THEME.text.fontSize.H2,
    paddingLeft: "7%",
    paddingRight: "7%",
  },
  bodyText: {
    color: THEME.text.color.secondary,
    fontSize: THEME.text.fontSize.modalBody,
  },
  headerText: {
    color: THEME.text.color.secondary,
    fontSize: THEME.text.fontSize.H3,
    fontWeight: "500",
  },
  modalErrorMessageText: {
    color: THEME.colors.danger,
    fontSize: THEME.text.fontSize.modalBody,
  },
  closeButton: { position: "absolute", top: 0, right: 0, padding: 10 },
  modalHeader: {
    alignItems: "flex-start",
    justifyContent: "center",
    flex: 0.3,
    width: "90%",
  },
  modalBody: {
    flex: 0.25,
    alignItems: "flex-start",
    justifyContent: "flex-start",
    width: "90%",
  },
  modalInputFields: {
    flex: 0.45,
    justifyContent: "flex-start",
    width: "90%",
  },
  modalErrorMessage: {
    flex: 0.1,
    alignItems: "center",
    justifyContent: "flex-start",
    width: "85%",
  },
  modalButtons: {
    flex: 0.1,
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "space-around",
    width: "70%",
    position: "absolute",
    bottom: 0,
    paddingBottom: "10%",
  },
  loadingIndicator: {
    position: "absolute",
    top: "50%",
    alignItems: "center",
  },
  modalSnackbar: {
    backgroundColor: THEME.colors.background,
  },
  snackbar: {
    backgroundColor: THEME.snackbar.color.background,
  },
});
