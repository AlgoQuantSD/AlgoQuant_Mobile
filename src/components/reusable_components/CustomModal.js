import React, { useState, useEffect } from "react";
import {
  View,
  Text,
  TouchableOpacity,
  StyleSheet,
  Keyboard,
  TouchableWithoutFeedback,
  ActivityIndicator,
} from "react-native";
import { TextInput, Button, Snackbar } from "react-native-paper";
import Modal from "react-native-modal";
import { Ionicons } from "@expo/vector-icons";
import { THEME } from "../../constants/Theme";
import {
  submitEditNameModal,
  submitDeleteAccountModal,
  submitResetBalanceModal,
  submitConnectAlpacaModal,
  submitDisconnectAlpacaModal,
  submitResetPasswordModal,
  submitUpdateEmailModalNewEmailStep,
  submitUpdateEmailConfirmNewEmailStep,
  submitUpdatePhoneModal,
  submitDeleteInvestorModal,
  submitStopJobModal,
} from "../../helpers/modalSubmitActions";
import { snackbarCleanUp } from "../../helpers/snackbarCleanup";
import TypewriterAnimatedText from "./TypewriterAnimatedText";

export default function CustomModal(props) {
  // Get the information for the modal and the functions to clear the information when we close the modal
  const {
    isModalVisible,
    setIsModalVisible,
    modalType,
    setModalType,
    modalTitle,
    setModalTitle,
    modalHeader,
    setModalHeader,
    modalBody,
    setModalBody,
    modalInputFields,
    setModalInputFields,
    modalButtons,
    setModalButtons,
    snackbarMessage,
    setSnackbarMessage,
    isSnackbarVisible,
    setIsSnackbarVisible,
    modalSnackbarMessage,
    setModalSnackbarMessage,
    isModalSnackbarVisible,
    setIsModalSnackbarVisible,
    modalErrorMessage,
    setModalErrorMessage,
    setShouldNavigateBack,
    jobID,
  } = props;

  // Close the modal and clear the modal information
  function handleModalClose() {
    setIsModalVisible(!isModalVisible);
    setModalType(null);
    setModalTitle(null);
    setModalHeader(null);
    setModalBody(null);
    if (setModalInputFields) {
      setModalInputFields(null);
    }
    setModalButtons(null);
    if (setInputValues) {
      setInputValues(null);
    }
    if (setIsModalSnackbarVisible) {
      setIsModalSnackbarVisible(false);
    }
    if (setModalSnackbarMessage) {
      setModalSnackbarMessage(null);
    }
    if (setIsSensitiveTextHidden) {
      setIsSensitiveTextHidden(true);
    }
  }

  // Perform the appropriate action upon submitting based on which modal is open
  function handleSubmit() {
    let submitProps = {
      inputValues,
      modalType,
      setInputValues,
      modalInputFields,
      isModalVisible,
      setIsModalVisible,
      setModalType,
      setModalTitle,
      setModalHeader,
      setModalBody,
      setModalInputFields,
      setModalButtons,
      isSnackbarVisible,
      setIsSnackbarVisible,
      snackbarMessage,
      setSnackbarMessage,
      modalSnackbarMessage,
      setModalSnackbarMessage,
      isModalSnackbarVisible,
      setIsModalSnackbarVisible,
      modalErrorMessage,
      setModalErrorMessage,
      isLoading,
      setIsLoading,
      setShouldNavigateBack,
      jobID,
    };
    switch (modalType) {
      case "EDIT_NAME":
        submitEditNameModal(submitProps);
        break;
      case "RESET_ALPACA_BALANCE":
      case "RESET_SIMULATED_BALANCE":
        submitResetBalanceModal(submitProps);
        break;
      case "CONNECT_ALPACA":
        submitConnectAlpacaModal(submitProps);
        break;
      case "DISCONNECT_ALPACA":
        submitDisconnectAlpacaModal(submitProps);
        break;
      case "RESET_PASSWORD":
        submitResetPasswordModal(submitProps);
        break;
      case "UPDATE_EMAIL_NEW_EMAIL_STEP":
        submitUpdateEmailModalNewEmailStep(submitProps);
        break;
      case "UPDATE_EMAIL_NEW_EMAIL_CONFIRM_STEP":
        submitUpdateEmailConfirmNewEmailStep(submitProps);
        break;
      case "UPDATE_PHONE":
        submitUpdatePhoneModal(submitProps);
        break;
      case "DELETE_ACCOUNT":
        submitDeleteAccountModal(submitProps);
        console.log("Account deleted!");
        break;
      case "DELETE_INVESTOR":
        console.log("Investor Deleted");
        submitDeleteInvestorModal(submitProps);
        break;
      case "STOP_JOB":
        console.log("Job Stopped");
        submitStopJobModal(submitProps);
        break;
      default:
    }
  }

  // Keep track of what is being typed into the input fields if we have any in our modal
  const [inputValues, setInputValues] = useState(
    Array(modalInputFields?.length).fill("")
  );
  const [isLoading, setIsLoading] = useState(false);

  const [isSensitiveTextHidden, setIsSensitiveTextHidden] = useState(true);
  // Refresh the component to get the correct amount of input values
  useEffect(() => {
    if (modalInputFields) {
      setInputValues(Array(modalInputFields.length).fill(""));
    }
  }, [modalInputFields]);

  // Check whether certain pieces of the modal should be rendered if so we render it
  // In some cases such as input fields and buttons we must loop through and render them because the number of input fields and buttons in a modal may vary
  return (
    <TouchableWithoutFeedback onPress={() => Keyboard.dismiss()}>
      <Modal
        isVisible={isModalVisible}
        onBackdropPress={() => setIsModalVisible(false)}
        style={styles.modalContainer}
      >
        {isLoading ? (
          <View style={styles.loadingIndicator}>
            <ActivityIndicator
              size="large"
              color={THEME.loadingIndicator.color}
            />
          </View>
        ) : null}
        <View style={styles.headerContainer}>
          <Text style={styles.titleText}>{modalTitle}</Text>
          <TouchableOpacity
            style={styles.closeButton}
            onPress={handleModalClose}
          >
            <Ionicons name="close" size={16} color="white" />
          </TouchableOpacity>
        </View>
        <View style={styles.contentContainer}>
          {modalHeader ? (
            <View style={styles.modalHeader} testID="modal-header">
              <Text style={styles.headerText}>{modalHeader}</Text>
            </View>
          ) : null}
          {modalBody ? (
            <View style={styles.modalBody} testID="modal-body">
              <Text style={styles.bodyText}>{modalBody}</Text>
            </View>
          ) : null}
          {modalInputFields ? (
            <View style={styles.modalInputFields} testID="modal-input-fields">
              {modalInputFields.map((item, index) => (
                <View
                  key={item.key}
                  style={{
                    flexDirection: "row",
                    width: "100%",
                  }}
                >
                  <TextInput
                    label={item.label}
                    secureTextEntry={
                      modalType === "RESET_PASSWORD"
                        ? isSensitiveTextHidden
                        : false
                    }
                    defaultValue={item.defaultValue}
                    onChangeText={(text) => {
                      const newInputValues = [...inputValues];
                      newInputValues[index] = text;
                      setInputValues(newInputValues);
                    }}
                    selectionColor={THEME.colors.background}
                    underlineColor={THEME.colors.background}
                    activeUnderlineColor={THEME.colors.background}
                    outlineColor={THEME.colors.background}
                    activeOutlineColor={THEME.colors.background}
                    textColor={THEME.colors.background}
                    placeholderTextColor={THEME.colors.background}
                    contentStyle={{ color: THEME.colors.background }}
                    style={{
                      backgroundColor: THEME.colors.transparent,
                      width: "100%",
                    }}
                  />
                  {modalType === "RESET_PASSWORD" ? (
                    <TouchableOpacity
                      style={{
                        alignSelf: "center",
                        position: "absolute",
                        left: "90%",
                      }}
                      onPress={() =>
                        setIsSensitiveTextHidden(!isSensitiveTextHidden)
                      }
                    >
                      <Ionicons
                        name="eye"
                        size={20}
                        color={THEME.colors.foreground}
                      />
                    </TouchableOpacity>
                  ) : null}
                </View>
              ))}
            </View>
          ) : null}
        </View>
        {modalErrorMessage ? (
          <View style={styles.modalErrorMessage}>
            <TypewriterAnimatedText
              text={modalErrorMessage}
              style={styles.modalErrorMessageText}
            />
          </View>
        ) : null}
        {modalButtons ? (
          <View style={styles.modalButtons}>
            {modalButtons.map((item, index) => (
              <TouchableOpacity key={item.key} style={styles.button}>
                <Button
                  onPress={
                    item.key === "SUBMIT_BUTTON"
                      ? handleSubmit
                      : handleModalClose
                  }
                  children={item.label}
                  buttonColor={item.buttonColor}
                  textColor={item.textColor}
                  style={{ width: 100 }}
                />
              </TouchableOpacity>
            ))}
          </View>
        ) : null}
        <View
          style={{
            position: "absolute",
            bottom: -150,
            width: "100%",
          }}
        >
          <Snackbar
            visible={isModalSnackbarVisible}
            onDismiss={() =>
              snackbarCleanUp(
                setIsModalSnackbarVisible,
                setModalSnackbarMessage
              )
            }
            duration={3500}
            action={{
              label: "Dismiss",
              textColor: THEME.text.color.secondary,
              onPress: () => {
                snackbarCleanUp(
                  setIsModalSnackbarVisible,
                  setModalSnackbarMessage
                );
              },
            }}
            style={styles.snackbar}
          >
            {modalSnackbarMessage}
          </Snackbar>
        </View>
      </Modal>
    </TouchableWithoutFeedback>
  );
}

const styles = StyleSheet.create({
  modalContainer: {
    justifyContent: "flex-start",
    alignItems: "center",
    backgroundColor: THEME.modal.color.background,
    flex: 0.75,
    marginTop: "50%",
    borderRadius: 10,
  },
  contentContainer: {
    flex: 0.6,
    alignItems: "center",
    justifyContent: "flex-start",
    width: "90%",
  },
  headerContainer: {
    position: "relative",
    flex: 0.15,
    flexDirection: "row",
    justifyContent: "center",
    alignItems: "center",
    width: "100%",
  },
  titleText: {
    color: THEME.text.color.secondary,
    fontSize: THEME.text.fontSize.H1,
  },
  bodyText: {
    color: THEME.text.color.secondary,
    fontSize: THEME.text.fontSizeModalBody,
  },
  headerText: {
    color: THEME.text.color.secondary,
    fontSize: THEME.text.fontSize.H3,
  },
  modalErrorMessageText: {
    color: THEME.colors.danger,
    fontSize: THEME.text.fontSize.modalBody,
  },
  closeButton: { position: "absolute", top: 0, right: 0, padding: 10 },
  modalHeader: {
    alignItems: "flex-start",
    justifyContent: "center",
    flex: 0.2,
    width: "90%",
  },
  modalBody: {
    flex: 0.25,
    alignItems: "flex-start",
    justifyContent: "flex-start",
    width: "90%",
  },
  modalInputFields: {
    flex: 0.45,
    justifyContent: "flex-start",
    width: "90%",
  },
  modalErrorMessage: {
    flex: 0.1,
    alignItems: "center",
    justifyContent: "flex-start",
    width: "85%",
  },
  modalButtons: {
    flex: 0.1,
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "space-around",
    width: "70%",
    position: "absolute",
    bottom: 0,
    paddingBottom: "10%",
  },
  loadingIndicator: {
    position: "absolute",
    top: "50%",
    alignItems: "center",
  },
  modalSnackbar: {
    backgroundColor: THEME.colors.background,
  },
  snackbar: {
    backgroundColor: THEME.snackbar.color.background,
  },
});
