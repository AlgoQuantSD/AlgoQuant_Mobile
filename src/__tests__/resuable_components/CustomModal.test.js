import React from "react";
import { render, screen } from "@testing-library/react-native";
import CustomModal from "../../components/reusable_components/CustomModal";
import { THEME } from "../../constants/Theme";
jest.useFakeTimers();

describe("CustomModal", () => {
  it("should render", () => {
    const tree = render(<CustomModal />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("renders the edit name modal correctly", () => {
    render(
      <CustomModal
        isModalVisible={true}
        setIsModalVisible={jest.fn()}
        modalType="EDIT_NAME"
        setModalType={jest.fn()}
        modalTitle="Edit Name"
        setModalTitle={jest.fn()}
        modalHeader={null}
        setModalHeader={jest.fn()}
        modalBody={null}
        setModalBody={jest.fn()}
        modalInputFields={[
          { label: "First Name", key: "FIRST_NAME_LABEL" },
          { label: "Last Name", key: "LAST_NAME_LABEL" },
        ]}
        setModalInputFields={jest.fn()}
        modalButtons={[
          {
            label: "Submit",
            buttonColor: THEME.colors.success,
            textColor: THEME.text.color,
            key: "SUBMIT_BUTTON",
          },
          {
            label: "Cancel",
            buttonColor: THEME.colors.danger,
            textColor: THEME.text.color,
            key: "CANCEL_BUTTON",
          },
        ]}
        setModalButtons={jest.fn()}
      />
    );
    const firstNameLabel = screen.getAllByText("First Name")[0];
    const lastNameLabel = screen.getAllByText("Last Name")[0];

    // Assert that all the parts of the modal that should be visible are visible
    expect(screen.getByText("Edit Name")).toBeTruthy();
    expect(firstNameLabel).toBeTruthy();
    expect(lastNameLabel).toBeTruthy();
    expect(screen.getByText("Submit")).toBeTruthy();
    expect(screen.getByText("Cancel")).toBeTruthy();

    // Assert that all the parts of the modal that should not be visible are not visible
    expect(screen.queryByTestId("modal-header")).toBeNull();
    expect(screen.queryByTestId("modal-body")).toBeNull();
  });

  it("does not render if isModalVisible is false", () => {
    render(
      <CustomModal
        isModalVisible={false}
        setIsModalVisible={jest.fn()}
        modalType="EDIT_NAME"
        setModalType={jest.fn()}
        modalTitle="Edit Name"
        setModalTitle={jest.fn()}
        modalHeader="Please enter your new name"
        setModalHeader={jest.fn()}
        modalBody={null}
        setModalBody={jest.fn()}
        modalInputFields={[
          { label: "First Name", key: "FIRST_NAME_LABEL" },
          { label: "Last Name", key: "LAST_NAME_LABEL" },
        ]}
        setModalInputFields={jest.fn()}
        modalButtons={[
          {
            label: "Submit",
            buttonColor: THEME.colors.success,
            textColor: THEME.text.color,
            key: "SUBMIT_BUTTON",
          },
          {
            label: "Cancel",
            buttonColor: THEME.colors.danger,
            textColor: THEME.text.color,
            key: "CANCEL_BUTTON",
          },
        ]}
        setModalButtons={jest.fn()}
      />
    );

    // Assert that nothing is visible
    expect(screen.queryByText("Edit Name")).toBeNull();
    expect(screen.queryByText("Please enter your new name")).toBeNull();
    expect(screen.queryByText("First Name")).toBeNull();
    expect(screen.queryByText("Last Name")).toBeNull();
    expect(screen.queryByText("Submit")).toBeNull();
    expect(screen.queryByText("Cancel")).toBeNull();
  });
});
