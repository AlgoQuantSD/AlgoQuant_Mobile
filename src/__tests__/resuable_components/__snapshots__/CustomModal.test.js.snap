// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomModal Should render 1`] = `
<Modal
  accessible={true}
  animationType="none"
  deviceHeight={null}
  deviceWidth={null}
  focusable={true}
  hardwareAccelerated={false}
  hideModalContentWhileAnimating={false}
  onBackdropPress={[Function]}
  onClick={[Function]}
  onModalHide={[Function]}
  onModalWillHide={[Function]}
  onModalWillShow={[Function]}
  onRequestClose={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  panResponderThreshold={4}
  scrollHorizontal={false}
  scrollOffset={0}
  scrollOffsetMax={0}
  scrollTo={null}
  statusBarTranslucent={false}
  supportedOrientations={
    Array [
      "portrait",
      "landscape",
    ]
  }
  swipeThreshold={100}
  transparent={true}
  visible={false}
/>
`;

exports[`CustomModal should render 1`] = `
<Modal
  accessible={true}
  animationType="none"
  deviceHeight={null}
  deviceWidth={null}
  focusable={true}
  hardwareAccelerated={false}
  hideModalContentWhileAnimating={false}
  onBackdropPress={[Function]}
  onClick={[Function]}
  onModalHide={[Function]}
  onModalWillHide={[Function]}
  onModalWillShow={[Function]}
  onRequestClose={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  panResponderThreshold={4}
  scrollHorizontal={false}
  scrollOffset={0}
  scrollOffsetMax={0}
  scrollTo={null}
  statusBarTranslucent={false}
  supportedOrientations={
    Array [
      "portrait",
      "landscape",
    ]
  }
  swipeThreshold={100}
  transparent={true}
  visible={false}
/>
`;
